### PR TITLE
Delete line if suggestion would replace it with an empty line

### LINKED ISSUE #120296: `cargo fix` on unused `use` statements leaves empty lines behind.
### Code

Run `cargo fix` on the following:
```Rust
use std::time::Duration;
use std::time::Instant;
use std::time::SystemTime;
```

### Current output

cargo fix will remove the unused `use` statements but will leave an empty line behind for each one. In this case that is 3 empty lines:
```



```

### Desired output

No empty lines left behind:
```
```

### Rationale and extra context

Leaving those empty lines around means they need to be manually cleaned up and I dont think there is ever a scenario where someone would want the empty line left behind.

I'm not sure if this is a cargo bug or a rustc bug.
I know rustc emits some kind of span to tell cargo what change needs to occur?
But maybe its cargo not interpreting the span correctly?

### Other cases

_No response_

### Rust Version

```Shell
It has been this way for as long as I can remember but:
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #120296

### CODE REVIEW

File: compiler/rustc_resolve/src/check_unused.rs [hunk 1]
```diff
@@ -373,7 +373,14 @@ impl Resolver<'_, '_> {
...
             );
 
-            let fix_msg = if fixes.len() == 1 && fixes[0].0 == unused.item_span {
+            let fix_msg = if let [(sp, _)] = fixes.as_mut_slice()
+                && *sp == unused.item_span
+            {
+                // If this is the only import on the line, remove the whole line
+                if let Ok(next) = tcx.sess.source_map().span_to_next_source(*sp) {
+                    let whitespace = next.chars().take_while(char::is_ascii_whitespace).count();
+                    *sp = sp.with_hi(sp.hi() + rustc_span::BytePos(whitespace as u32));
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/check_unused.rs:382]: `whitespace` here is counted in `char`s, not bytes.

Comment (petrochenkov) [reply compiler/rustc_resolve/src/check_unused.rs:382]: Are there any other suggestions that remove full lines?
There probably are, in that case would be better to share this logic with them.

Comment (clubby789) [reply compiler/rustc_resolve/src/check_unused.rs:382]: Given that the iterator takes only `ascii_whitespace`, I think it makes sense to treat the char count as byte count? I'll look for similar suggestions

**Review by estebank — COMMENTED** [review]
Although this is a way to address this specific issue, it might be a better idea to modify `rustc_errors::json::DiagnosticSpan::from_span_full` to have this logic of checking "does this span end on the last char of a line, followed by `\n`? if extend the span by one byte". Doing it there *doesn't* change how the diff is shown in diagnostics, but aids rustfix not have to care about this *and* doesn't require us to address these kind of spans in *every* suggestion we produce.

**Comment by clubby789** [comment]
Doing that globally seems to break a lot of `.fixed` tests:
```diff,rs
diff --git a/tests/ui/binop/false-binop-caused-by-missing-semi.fixed b/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
index b47372c9064..441fad80c8f 100644
--- a/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
+++ b/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
@@ -3,8 +3,7 @@ fn foo() {}
 fn main() {
     let mut y = 42;
     let x = &mut y;
-    foo();
-    *x = 0;  //~ ERROR invalid left-hand side of assignment
+    foo();    *x = 0;  //~ ERROR invalid left-hand side of assignment
     let _ = x;
     println!("{y}");
 }
diff --git a/tests/ui/block-result/consider-removing-last-semi.fixed b/tests/ui/block-result/consider-removing-last-semi.fixed
index 36a769fe529..931a5440b79 100644
--- a/tests/ui/block-result/consider-removing-last-semi.fixed
+++ b/tests/ui/block-result/consider-removing-last-semi.fixed
@@ -2,19 +2,16 @@
 
 pub fn f() -> String {  //~ ERROR mismatched types
     0u8;
-    "bla".to_string()
-}
+    "bla".to_string()}
 
 pub fn g() -> String {  //~ ERROR mismatched types
     "this won't work".to_string();
-    "removeme".to_string()
-}
+    "removeme".to_string()}
 
 pub fn macro_tests() -> u32 {  //~ ERROR mismatched types
     macro_rules! mac {
         () => (1);
     }
-    mac!()
-}
+    mac!()}```

**Comment by estebank** [comment]
@clubby789 what happens if you only do the span extension if the replacement is ""?

**Comment by clubby789** [comment]
Catches a lot more cases 👀 Good idea!

**Comment by clubby789** [comment]
Actually this could do with some improvement
@rustbot author

**Review by estebank — APPROVED** [review]
please rebase and I'll approve

**Pull request merged**